transport-classes-io_uring/src/main/java/io/netty/channel/uring/IOUringIoHandler.java

```diff
@@ -0,0 +1,440 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * {@link IoHandler} which is implemented in terms of the Linux-specific {@code io_uring} API.
+ */
+public final class IOUringIoHandler implements IoHandler {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(IOUringIoHandler.class);
+    private static final short RING_CLOSE = 1;
+
+    private final RingBuffer ringBuffer;
+    private final IntObjectMap<DefaultIoUringIoRegistration> registrations;
+    // The maximum number of bytes for an InetAddress / Inet6Address
+    private final byte[] inet4AddressArray = new byte[SockaddrIn.IPV4_ADDRESS_LENGTH];
+    private final byte[] inet6AddressArray = new byte[SockaddrIn.IPV6_ADDRESS_LENGTH];
+
+    private final AtomicBoolean eventfdAsyncNotify = new AtomicBoolean();
+    private final FileDescriptor eventfd;
+    private final long eventfdReadBuf;
+
+    private final CompletionCallback completionCallback = new CompletionCallback() {
+        @Override
+        public void handle(int res, int flags, int id, byte op, short data) {
+            IOUringIoHandler.this.handle(res, flags, id, op, data);
+        }
+    };
+
+    private long eventfdReadSubmitted;
+    private boolean eventFdClosing;
+    private volatile boolean shuttingDown;
+    private boolean closeCompleted;
+    private int nextRegistrationId = Integer.MIN_VALUE;
+
+    // these two ids are used internally any so can't be used by nextRegistrationId().
+    private static final int EVENTFD_ID = Integer.MAX_VALUE;
+    private static final int RINGFD_ID = EVENTFD_ID - 1;
+
+    IOUringIoHandler(RingBuffer ringBuffer) {
+        // Ensure that we load all native bits as otherwise it may fail when try to use native methods in IovArray
+        IOUring.ensureAvailability();
+        this.ringBuffer = requireNonNull(ringBuffer, "ringBuffer");
+        registrations = new IntObjectHashMap<>();
+        eventfd = Native.newBlockingEventFd();
+        eventfdReadBuf = PlatformDependent.allocateMemory(8);
+    }
+
+    @Override
+    public int run(IoExecutionContext context) {
+        SubmissionQueue submissionQueue = ringBuffer.ioUringSubmissionQueue();
+        CompletionQueue completionQueue = ringBuffer.ioUringCompletionQueue();
+        if (!completionQueue.hasCompletions() && context.canBlock()) {
+            if (eventfdReadSubmitted == 0) {
+                submitEventFdRead();
+            }
+            if (context.deadlineNanos() != -1) {
+                submitTimeout(context);
+            }
+            submissionQueue.submitAndWait();
+        } else {
+            submissionQueue.submit();
+        }
+        return completionQueue.process(completionCallback);
```

Comment:
Instead of `completionCallback` we could pass a method reference to `handle`.
Then we wouldn't need the field and anonymous class for `completionCallback`.